api/conf/lifecycle_handler.py

```diff
@@ -6,31 +6,41 @@
 
 from django.core.handlers.asgi import ASGIHandler
 
+import sentry_sdk
+
 
 parent_logger = logging.getLogger(__name__)
 
 
-class OpenverseASGIHandler(ASGIHandler):
+class ASGILifecycleHandler:
     """
-    Extend default ASGIHandler to implement lifetime hooks.
+    Handle ASGI lifecycle messages.
+
+    Django's ASGIHandler does not handle these messages,
+    so we have to implement it ourselves. Only shutdown handlers
+    are currently supported.
```

Comment:
I read in the Django bug tracker that ASGI lifecycle wasn't supported. Somehow I didn't come across this project, though, which I wonder if we should use instead of our custom implementation: https://github.com/illagrenan/django-asgi-lifespan

Reply:
Update: yeah, I'm going to go ahead and change this PR to use that package so that we don't have to do a refactor later to remove our custom implementation.

Reply:
Ah! I remember why I originally did not use that package: it only supports uvicorn directly, but not under gunicorn workers. That's not an issue for us anymore, so we can use it.

It doesn't have a lot of activity, but it has comprehensive unit tests and by using Django signals, gets around the need to implement custom weak ref handling (Django signal handler refs are weak by default).

Reply:
Okay, that's done in the latest commit: https://github.com/WordPress/openverse/pull/3024/commits/e4dc958b8e14ebcf4a4021c7a632b6769ebf41b4

It actually was a pretty nice way to clean things up. Rather than registering each session's close and needing to manage it, we can benefit from the `_SESSIONS` gathering point and register a single signal handler that closes all the sessions.

I had to refactor the test fixture as well, and almost thought it was finally going to break completely, until I realised I'd come across a bug in the upstream library, for which I've submitted a patch: https://github.com/illagrenan/django-asgi-lifespan/pull/80

It doesn't currently affect us, and shouldn't in the foreseeable future, because we don't need the startup lifecycle event (and worst-case scenario, we can still call the startup event without awaiting it and just tolerate a log at the end of the tests complaining that it was never awaited).

In any case, I think the end result is a lot cleaner and easier to follow, and easier to extend into the future.